File: palindrome_rectangle.py

class CharMatrix:

    def __init__(self, input_matrix):

        if len(input_matrix) == 0:
            raise ValueError("input_matrix is empty")

        lines = input_matrix.split('\n')

        self.width = len(lines[0])
        self.height = len(lines)

        for line in lines:
            if len(line) != self.width:
                raise ValueError("The char matrix must have lines with equal sizes")

        self.input_matrix = input_matrix.replace('\n', '')

    def at(self, i, j):

        if i >= self.width or j >= self.height:
            raise IndexError("index out of bounds")

        # print("CharMatrix {i},{j} = {c}".format(i=i, j=j, c=self.input_matrix[i+j*self.width]))
        return self.input_matrix[i + j * self.width]

    def __str__(self):
        return "CharMatrix({width}, {height})".format(width=self.width, height=self.height)


class RectanglePalindromeAnalyser:

    def __init__(self, char_matrix):

        if not isinstance(char_matrix, CharMatrix):
            raise ValueError("char_matrix must be of type CharMatrix")

        if char_matrix.width < 2 or char_matrix.height < 2:
            raise ValueError("char_matrix must be at least 2x2")

        self.char_matrix = char_matrix
        self.num_op = 0

    def find_brute_force(self):

        idx_i = -1
        idx_j = -1
        rect_width = -1
        rect_height = -1

        for rw in range(2, self.char_matrix.width):
            for rh in range(2, self.char_matrix.height):

                for i in range(self.char_matrix.width - rw):
                    for j in range(self.char_matrix.height - rh):

                        if self.__is_rect_palindrome(i, j, rw, rh):
                            idx_i = i
                            idx_j = j
                            rect_width = rw
                            rect_height = rh

        return idx_i, idx_j, rect_width, rect_height, self.num_op

    def __is_rect_palindrome(self, start_i, start_j, width, height):
        is_pal_hor1 = self.__is_palindrome(start_i, start_j, width, 'hor')
        is_pal_hor2 = self.__is_palindrome(start_i, start_j + height - 1, width, 'hor')
        is_pal_ver1 = self.__is_palindrome(start_i, start_j, height, 'ver')
        is_pal_ver2 = self.__is_palindrome(start_i + width - 1, start_j, height, 'ver')

        # print(is_pal_hor1, is_pal_hor2, is_pal_ver1, is_pal_ver2)

        return is_pal_hor1 and is_pal_hor2 and is_pal_ver1 and is_pal_ver2

    def __is_palindrome(self, start_i, start_j, size, direction):
        b_is_palindrome = True

        if direction == 'ver':
            for i in range(int(size / 2)):
                idx = start_j + i
                ridx = size - i + start_j - 1

                char1 = self.char_matrix.at(start_i, idx)
                char2 = self.char_matrix.at(start_i, ridx)

                self.num_op += 1

                # print(char1, char2)

                b_is_palindrome &= char1 == char2

        if direction == 'hor':
            for j in range(int(size / 2)):
                idx = start_i + j
                ridx = size - j + start_i - 1

                char1 = self.char_matrix.at(idx, start_j)
                char2 = self.char_matrix.at(ridx, start_j)

                self.num_op += 1

                # print(char1, char2)

                b_is_palindrome &= char1 == char2

        return b_is_palindrome


def is_palindrome(word):
    """

    Return if a word is a palindrome.
    Complexity O(len(word))

    :param word:
    :return:
    """
    b_is_palindrome = True

    for i in range(int(len(word)/2)):
        print(word[i], word[-i])
        b_is_palindrome &= word[i] == word[len(word) - i - 1]

    return b_is_palindrome


def main():
    input_rectangle = 'QWSDOP\n' \
                      'QWSSWP\n' \
                      'QDSDOP\n' \
                      'QDSDOP\n' \
                      'QWRRWP\n' \
                      'QWSDOP'

    char_matrix = CharMatrix(input_rectangle)
    palindrome_analyser = RectanglePalindromeAnalyser(char_matrix)
    print(palindrome_analyser.find_brute_force())

if __name__ == '__main__':
    main()

File: PalindromeRectangle.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <iterator>

using namespace std;


class CharMatrix
{
public:
	int Width;
	int Height;
	string Matrix;

	CharMatrix(const string& InputMatrix)
	{
		if (InputMatrix.empty())
		{
			return;
		}

		istringstream MatrixStream{ InputMatrix };

		vector<string> Tokens{ istream_iterator<string>{ MatrixStream }, istream_iterator<string>{} };

		Width = Tokens.front().size();
		Height = Tokens.size();

		for (auto Token : Tokens)
		{
			if (Token.size() != Width)
			{
				return;
			}

			Matrix += Token;
		}
	}

	char operator()(int i, int j) const
	{
		return Matrix.at(i + j * Width);
	}

	friend ostream& operator<< (ostream& out, const CharMatrix& Matrix)
	{
		out << "CharMatrix(" << Matrix.Width << ", " << Matrix.Height << ") = " << Matrix.Matrix;
		return out;
	}
};


class RectanglePalindromeAnalyser
{
public:
	CharMatrix& InputMatrix;
	int NumOp = 0;

	enum class Direction
	{
		Horizontal,
		Vertical
	};

	RectanglePalindromeAnalyser(CharMatrix& Matrix) :
		InputMatrix(Matrix)
	{
		if (InputMatrix.Width < 2 || InputMatrix.Height < 2)
		{
			return;
		}
	}

	void FindBruteForce()
	{
		int Idx_i = -1;
		int Idx_j = -1;
		int RectWidth = -1;
		int RectHeight = -1;

		for (int RW = 2; RW < InputMatrix.Width; ++RW)
		{
			for (int RH = 2; RH < InputMatrix.Height; ++RH)
			{
				for (int i = 0; i < InputMatrix.Width - RW; ++i)
				{
					for (int j = 0; j < InputMatrix.Height - RH; ++j)
					{
						if (IsRectPalindrome(i, j, RW, RH))
						{
							Idx_i = i;
							Idx_j = j;
							RectWidth = RW;
							RectHeight = RH;
						}
					}
				}
			}
		}

		cout << Idx_i << ", " << Idx_j << ", " << RectWidth << ", " << RectHeight << ", " << NumOp << endl;
	}

	bool IsRectPalindrome(int Start_i, int Start_j, int Width, int Height)
	{
		bool bIsPalHor1 = IsPalindrome(Start_i, Start_j, Width, Direction::Horizontal);
		bool bIsPalHor2 = IsPalindrome(Start_i, Start_j + Height - 1, Width, Direction::Horizontal);
		bool bIsPalVer1 = IsPalindrome(Start_i, Start_j, Height, Direction::Vertical);
		bool bIsPalVer2 = IsPalindrome(Start_i + Width - 1, Start_j, Height, Direction::Vertical);

		return bIsPalHor1 && bIsPalHor2 && bIsPalVer1 && bIsPalVer2;
	}

	bool IsPalindrome(int Start_i, int Start_j, int Size, Direction Dir)
	{
		bool bIsPalindrome = true;

		if (Dir == Direction::Vertical)
		{
			for (int i = 0; i < Size / 2; ++i)
			{
				int Idx = Start_j + i;
				int RIdx = Size - i + Start_j - 1;

				char Char1 = InputMatrix(Start_i, Idx);
				char Char2 = InputMatrix(Start_i, RIdx);

				NumOp++;

				bIsPalindrome &= Char1 == Char2;
			}		
		}
		else if (Dir == Direction::Horizontal)
		{
			for (int j = 0; j < Size / 2; ++j)
			{
				int Idx = Start_i + j;
				int RIdx = Size - j + Start_i - 1;

				char Char1 = InputMatrix(Idx, Start_j);
				char Char2 = InputMatrix(RIdx, Start_j);

				NumOp++;

				bIsPalindrome &= Char1 == Char2;
			}
		}

		return bIsPalindrome;
	}
};



int main()
{
	string InputRectangle = "QWSDOP\n"
							"QWSSWP\n"
							"QDSDOP\n"
							"QDSDOP\n"
							"QWRRWP\n"
							"QWSDOP";

	CharMatrix InputMatrix{ InputRectangle };
	RectanglePalindromeAnalyser PalAnalyser{ InputMatrix };

	PalAnalyser.FindBruteForce();
	

	return 0;
}